File: 323. Number of Connected Components in an Undirected Graph/solution.py
class Solution(object):
    def countComponents(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: int
        """
        # create graph
        graph = {}
        for v in range(n):
            graph[v] = set()
        for v, w in edges:
            graph[v].add(w)
            graph[w].add(v)
                
        # perform dfs and count number of components
        count = 0
        visited = [False]*n
        for v in range(n):
            if not visited[v]:
                self.dfs(graph, visited, v)
                count += 1
                
        return count
    
    def dfs(self, graph, visited, v):
        if visited[v]:
            return
        visited[v] = True
        
        for w in graph[v]:
            self.dfs(graph, visited, w)

File: 323. Number of Connected Components in an Undirected Graph/solution.java
class Solution {
    private int[] parent;
    public int find(int x) {
        if (parent[x] == x) {
            return x;
        } else {
            parent[x] = find(parent[x]);
            return parent[x];
        }
    }
    
    public void union(int x, int y) {
        int parentX = find(x);
        int parentY = find(y);
        if (parentX != parentY) {
            parent[parentY] = parentX;
        }
    }
    
    public int countComponents(int n, int[][] edges) {
        // init
        parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
        
        // union
        int count = 0;
        for (int[] edge: edges) {
            union(edge[0], edge[1]);
        }
        
        // count
        for (int i = 0; i < n; i++) {
            if (parent[i] == i){
                count++;
            }
        }
        
        return count;
    }
}
